### ISSUE TITLE
[aws-http4s] Credential provider chain swallows errors, probably works incorrectly on ECS/EC2

### ISSUE THREAD

**Issue by keynmol** [disneystreaming/smithy4s#1464]
(follow up from DM with @Baccata)

This one is difficult to reproduce, but the issue can be summarised in this screenshot:

![CleanShot 2024-04-02 at 11 53 23](https://github.com/disneystreaming/smithy4s/assets/1052965/7c5281a1-4cb1-4695-83bf-ec0a3cd9fd65)

In short, using the [default](https://github.com/disneystreaming/smithy4s/blob/series/0.18/modules/aws-http4s/src/smithy4s/aws/AwsCredentialsProvider.scala#L70) credentials provider seems to be brittle - even though the ECS endpoint env variable is present, for some reason getting credentials through that fails, and the error that bubbles up is for the credentials file.

My hunch is that actual source of the error is timeouts (e.g. during task start up, 1 second might not be enough) and lack of retries. My current workaround is to manually create a provider with desired logic:

```scala
  def awsEnvironment(
      httpClient: org.http4s.client.Client[IO]
  ): Resource[IO, AwsEnvironment[IO]] =
    std
      .Env[IO]
      .get("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
      .toResource
      .flatMap:
        case None =>
          AwsEnvironment.default[IO](httpClient, AWS_REGION)

        // Below is a hack to work around a bug in credentials loading
        // chain in smithy4s - we manually prioritise ECS credentials
        case Some(_) =>
          val provider = new AwsCredentialsProvider[IO]
          provider
            .refreshing(
              provider
                .fromECS(httpClient, 10.second)
                .onError(exc =>
                  scribe.cats.io
                    .error("Failed to get credentials from ECS endpoint", exc)
                )
            )
            .map: cred =>
              AwsEnvironment.make(
                httpClient,
                IO.pure(AWS_REGION),
                cred,
                IO.realTime.map(_.toSeconds).map(Timestamp(_, 0))
              )

```

But I think this whole credentials chain thing should be redesigned to support

1. Enabling/disabling providers
2. Retries for network-based providers
3. Error accumulation 
4. Per-provider error summary when everything fails

**Comment by kubukoz** [comment]
I agree, we should have some retries and more visibility - `orElse` just swallows earlier errors which makes it difficult to understand how you're landing in the "file not found" case.

**Comment by Baccata** [comment]
I don't disagree with the need to make it better, my only request is that if loggers need to be invoked, pass it as a `LogRecord => F[Unit]` function as opposed to adding a dependency to some logging library. 

To clarify : I'm happy to read PRs (though I may have to be pestered at for it), but I don't have the bandwidth to take this on.
